task/control-plane-agent/src/dump.rs

```diff
@@ -0,0 +1,276 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+use core::mem::size_of;
+use dump_agent_api::DumpAgent;
+use gateway_messages::{
+    DumpCompression, DumpError, DumpSegment, DumpTask, SpError,
+};
+use userlib::{task_slot, UnwrapLite};
+use zerocopy::{AsBytes, FromBytes};
+
+task_slot!(DUMP_AGENT, dump_agent);
+
+const DUMP_TASK_SIZE: u32 = size_of::<humpty::DumpTask>() as u32;
+const HEADER_SIZE: u32 = size_of::<humpty::DumpAreaHeader>() as u32;
+const SEGMENT_DATA_SIZE: u32 = size_of::<humpty::DumpSegmentData>() as u32;
+
+#[derive(Copy, Clone)]
+struct ClientDumpState {
+    /// Key used to distinguish between clients (should be random)
+    key: [u8; 16],
+
+    /// Expected sequence number
+    expected_seq: u32,
+
+    /// Current position, associated with `expected_seq`
+    current_pos: Position,
+
+    /// Previous position, associated with `expected_seq - 1`
+    prev_pos: Option<Position>,
+}
+
+#[derive(Copy, Clone)]
+struct Position {
+    /// Index of the dump area being read
+    area_index: u8,
+
+    /// Offset into the dump area
+    ///
+    /// This should point to a `DumpSegmentData` header, followed by compressed
+    /// data.
+    offset: u32,
+}
+
+const MAX_DUMP_CLIENTS: usize = 4;
+pub struct DumpState {
+    agent: DumpAgent,
+    clients: [Option<ClientDumpState>; MAX_DUMP_CLIENTS],
+}
+
+impl DumpState {
+    pub fn new() -> Self {
+        DumpState {
+            clients: [None; MAX_DUMP_CLIENTS],
+            agent: DumpAgent::from(DUMP_AGENT.get_task_id()),
+        }
+    }
+
+    pub(crate) fn get_task_dump_count(&mut self) -> Result<u32, SpError> {
+        let mut count = 0;
+        for index in 0.. {
+            // XXX accidentally quadratic!
```

Comment:
I... think it's not accidental? ;) Worth turning this into a comment explaining that it's quadratic and not a problem or deleting it? (As it stands, it reads like a note to oneself.)